commit 21905912cbb2e977fb827b10a65ab0e6c772fe08 (refs/changes/91/691/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2016-10-14T12:20:22-07:00 (3 years ago)
    
    Revert "DOCKER-948 docker build fails with 'Build failed: problem starting container' error"
    
    This reverts commit 6a383d4960e4219a15ff0af00dfe2d6f5127de82.

diff --git a/lib/backends/sdc/build.js b/lib/backends/sdc/build.js
index cce4519..2253581 100644
--- a/lib/backends/sdc/build.js
+++ b/lib/backends/sdc/build.js
@@ -844,8 +844,11 @@ function runBuildCommand(opts, callback) {
     };
 
     /**
-     * The way `docker run` works is that an `attach` must be made first, and
-     * then the `start` call, and finally the `wait` call.
+     * This flow is a little strange, so an explanation is warranted:
+     * - attachContainer launches and then it waits for the container to start
+     *   before returning via the callback.
+     * - we synchronously call startContainer and thus this will allow
+     *   attachContainer to return, after which we call waitContainer.
      */
     req.backend.attachContainer({
         account: req.account,
@@ -863,35 +866,32 @@ function runBuildCommand(opts, callback) {
             callback(err);
             return;
         }
-
-        log.debug('runBuildCommand: startContainer');
-        req.backend.startContainer({
+        // Get return code from the run command.
+        req.backend.waitContainer({
             account: req.account,
             app: req.app,
             log: log,
             req_id: opts.req_id,
             vm: opts.vm
-        }, function (startErr) {
-            log.debug('runBuildCommand: startContainer finished, err: %j',
-                startErr);
-            if (startErr) {
-                callback(startErr);
-                return;
-            }
-
-            // Get return code from the run command.
-            req.backend.waitContainer({
-                account: req.account,
-                app: req.app,
-                log: log,
-                req_id: opts.req_id,
-                vm: opts.vm
-            }, function (waitErr, exitCode) {
-                log.debug('runBuildCommand: container exit code: %j', exitCode);
-                callback(waitErr, { exitCode: exitCode });
-            });
+        }, function (waitErr, exitCode) {
+            log.debug('runBuildCommand: container exit code: %j', exitCode);
+            callback(waitErr, { exitCode: exitCode });
         });
     });
+
+    log.debug('runBuildCommand: startContainer');
+    req.backend.startContainer({
+        account: req.account,
+        app: req.app,
+        log: log,
+        req_id: opts.req_id,
+        vm: opts.vm
+    }, function (err) {
+        log.debug('runBuildCommand: startContainer finished, err: %j', err);
+        if (err) {
+            callback(err);
+        }
+    });
 }
 
 
